webhook/test/WebhookTest.groovy

```diff
@@ -0,0 +1,756 @@
+import spock.lang.Specification
+import groovy.json.JsonSlurper
+import org.jfrog.artifactory.client.model.repository.settings.impl.MavenRepositorySettingsImpl
+import org.jfrog.artifactory.client.model.repository.settings.impl.DockerRepositorySettingsImpl
+import org.jfrog.artifactory.client.ArtifactoryClientBuilder
+import org.jfrog.lilypad.Control
+
+/**
+ * Tests for Webhook
+ *
+ *
+ * How the tests work:
+ * Generally the tests work by uploading a secondary plugin into Artifactory. This plugin adds REST API points for
+ * receiving events and storing the data. In other words, this secondary plugin is the target of events triggered by
+ * the webhook. A typical test will:
+ *
+ * 1. Clear the secondary plugins buffer
+ * 2. Upload a new webhook configuration file (with the event(s) we are testing)
+ * 3. Reload the config
+ * 4. Trigger the event
+ * 5.
+ *
+ */
+class WebhookTest extends Specification {
+    // Artifactory
+    static def BASE_URL = 'http://localhost:8088/artifactory'
+    static def BASE_PORT = 8088
+    static def RELOAD_PLUGINS_URL = 'api/plugins/reload'
+    static def BUILD_UPLOAD_URL = 'api/build'
+    static def MAVEN_REPO1_NAME = "maven-local"
+    static def MAVEN_REPO2_NAME = "maven-local2"
+    static def DOCKER_REPO_NAME = "docker-local"
+
+    // Webhook
+    static def WEBHOOK_PATH = '/var/opt/jfrog/artifactory/etc/plugins/webhook.groovy'
+    static def WEBHOOK_CONFIG_PATH = '/var/opt/jfrog/artifactory/etc/plugins/webhook.config.json'
+    static def WEBHOOK_INFO_URL = 'api/plugins/execute/webhookInfo'
+    static def WEBHOOK_PING_URL = 'api/plugins/execute/pingWebhook'
+    static def WEBHOOK_RELOAD_URL = 'api/plugins/execute/webhookReload'
+    // Loopback
+    static def WEBHOOK_LOOPBACK_PATH = '/var/opt/jfrog/artifactory/etc/plugins/webhookLoopback.groovy'
```

Comment:
The plugin code is written into the test itself. It is in the variable WEBHOOK_LOOPBACK_CODE and the test uses Lilypad to write it to the test Artifactory. It is done this way to avoid the user accidentally writing it to their own plugin dir.

Reply:
Sorry, I did not see it.